### PR TITLE
Add PyTorch AMP check

### LINKED ISSUE #7908: NaN tensor values problem for GTX16xx users (no problem on other devices)
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training, Validation

### Bug

I used [yolov5 ](https://github.com/ultralytics/yolov5)to test with the demo dataset (coco128) and found that box and obj are nan. Also, there are no detections appear on validation images. This only happens on GTX1660ti devices (GPU mode), when I use CPU or use Google colab(Tesla K80) / RTX2070  for training, everything works fine.
image

### Environment

- Windows 10 10.0.19044.1706
- YOLOv5-6.1 (version 6.1)
- Nvidia GTX 1660 TI, 6 GB
- Python3.9
- cudatoolkit-11.3.1
- pytorch-1.11.0-py3.9_cuda11.3_cudnn8_0
- (also tried pytorch-1.11.0-py3.9_cuda11.5_cudnn8_0)
- (with dependencies installed correctly)

### Minimal Reproducible Example

The command used for training is
`python train.py`

### Additional

There are issues here also discussing the same problem.

- https://github.com/pytorch/pytorch/issues/58123
- https://github.com/openai/glide-text2im/issues/31
- https://discuss.pytorch.org/t/half-precision-convolution-cause-nan-in-forward-pass/117358/3
- https://github.com/pytorch/pytorch/issues/69449
- https://github.com/ultralytics/yolov5/issues/5815

However, I have tried pytorch with cuda version 11.5 (whose cudnn version is 8.3.0>8.2.2) and I also tried **downloading cuDNN from nvidia and copy/paste the dll files into the relevant folder in torch/lib** , the problem still can not be solved.

Another workaround is to downgrade to pytorch with cuda version 10.2(tested and it works), but this is currently not feasible as CUDA-10.2 PyTorch builds are no longer available for Windows.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by glenn-jocher**
Help identify AMP issues before training starts. Issue raised in https://github.com/ultralytics/yolov5/issues/7908

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced AutoShape initialization with verbosity control and improved Automatic Mixed Precision (AMP) support in training.

### 📊 Key Changes
- Added `verbose` flag to `AutoShape` class for optional logging.
- Introduced `check_amp` function to verify AMP compatibility and functionality.
- Replaced direct amp import with `torch.cuda.amp` for AMP contexts.
- Adjusted model training to use AMP based on model compatibility check.
- Removed redundant import statements.

### 🎯 Purpose & Impact
- **Verbosity Control**: Allows quieter model initialization, reducing console clutter for users.
- **AMP Verification**: Ensures AMP works correctly with the model, which can help avoid training issues and support debugging.
- **AMP Utilization**: More precise handling of when to enable AMP, potentially improving training speed and memory efficiency.
- **Code Cleanup**: Streamlined codebase for better maintainability and clarity, yielding an easier-to-understand setup for developers and users alike. 🧹💻

🚀 For users, expect potentially faster and more efficient model training with the added comfort of toggling informational messages. For developers, this is a step towards a cleaner and more robust codebase.

### CODE REVIEW

File: utils/general.py [hunk 1]
```diff
@@ -505,6 +507,30 @@ def check_dataset(data, autodownload=True):
...
+
+    help_url = 'https://github.com/ultralytics/yolov5/issues/7908'
+    prefix = colorstr('AMP:')
+    device = next(model.parameters()).device  # get model device
+    if device.type == 'cpu':
+        return False
+    m = AutoShape(model, verbose=False)  # model
+    im = cv2.imread(ROOT / 'data' / 'images' / 'bus.jpg')[..., ::-1]  # OpenCV image (BGR to RGB)
+    y = m(im).xyxy[0]
+    with torch.cuda.amp.autocast(enabled=True):
```
Comment (YipKo) [inline utils/general.py]: add
`m.amp = True`


image

Comment (glenn-jocher) [reply utils/general.py]: Done!

**Comment by glenn-jocher** [comment]
@MarkDeia I've applied the changes. Can you check again and make sure it fails on your system?

**Comment by YipKo** [comment]
> @MarkDeia I've applied the changes. Can you check again and make sure it fails on your system?

Sure, I've run it locally with the cuda11 version of pytorch and it shows that the check failed, on colab it says that the check passed.Seems it works well.
image

**Review by YipKo — DISMISSED** [review]

**Comment by glenn-jocher** [comment]
@MarkDeia really strange. Are you sure the issue originates in this PR, i.e. PR fails with torchvision==0.11.2 but master is ok?

**Comment by YipKo** [comment]
> @MarkDeia really strange. Are you sure the issue originates in this PR, i.e. PR fails with torchvision==0.11.2 but master is ok?

@glenn-jocher I think there is something wrong with my local environment as master failed to run too.
I have just found a solution to this problem, it's because the version of pillow installed automatically by pip is too high.

**Comment by glenn-jocher** [comment]
@MarkDeia PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Pull request merged**